drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/rulesengine/MemoryMonitorUtil.java

```diff
@@ -0,0 +1,70 @@
+package org.drools.ansible.rulebook.integration.api.rulesengine;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemoryMonitorUtil {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MemoryMonitorUtil.class.getName());
+
+    public static final String MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY = "drools.memory.occupation.percentage.threshold";
+    private static final int DEFAULT_MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD = 90;
+    private static final int MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD;
+
+    // check memory per configured number of events are consumed
+    public static final String MEMORY_CHECK_EVENT_COUNT_THRESHOLD_PROPERTY = "drools.memory.check.event.count.threshold";
+    private static final int DEFAULT_MEMORY_CHECK_EVENT_COUNT_THRESHOLD = 100;
+    private static final int MEMORY_CHECK_EVENT_COUNT_THRESHOLD;
+
+    private static AtomicInteger eventCount = new AtomicInteger(0);
+
+    static {
+        String memoryThresholdEnvValue = System.getenv("DROOLS_MEMORY_THRESHOLD");
+        if (memoryThresholdEnvValue != null && !memoryThresholdEnvValue.isEmpty()) {
+            // Environment variable takes precedence over system property
+            System.setProperty(MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY, memoryThresholdEnvValue);
+        }
+        MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD = Integer.getInteger(MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY, DEFAULT_MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD); // percentage
+        LOG.info("Memory occupation threshold set to {}%", MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD);
+
+        String eventCountThresholdEnvValue = System.getenv("DROOLS_MEMORY_CHECK_EVENT_COUNT_THRESHOLD");
+        if (eventCountThresholdEnvValue != null && !eventCountThresholdEnvValue.isEmpty()) {
+            // Environment variable takes precedence over system property
+            System.setProperty(MEMORY_CHECK_EVENT_COUNT_THRESHOLD_PROPERTY, eventCountThresholdEnvValue);
+        }
+        MEMORY_CHECK_EVENT_COUNT_THRESHOLD = Integer.getInteger(MEMORY_CHECK_EVENT_COUNT_THRESHOLD_PROPERTY, DEFAULT_MEMORY_CHECK_EVENT_COUNT_THRESHOLD); // number of events
+        LOG.info("Memory check event count threshold set to {}", MEMORY_CHECK_EVENT_COUNT_THRESHOLD);
+    }
+
+    private MemoryMonitorUtil() {
+        // do not instantiate
+    }
+
+    public static void checkMemoryOccupation() {
+        // check memory every N events to avoid performance overhead
+        if (eventCount.incrementAndGet() >= MEMORY_CHECK_EVENT_COUNT_THRESHOLD) {
+            doCheckMemoryOccupation();
+            eventCount.set(0);
+        }
+    }
```

Comment:
@mariofusco Now check memory every N events to avoid the overhead. Default is `100`.

Btw, you wrote `(and running the gc) `, but isn't `System.gc()` per 100 events too frequent? Rather, shall I call `System.gc()` only when memoryOccupationPercentage is larger than 60%? (maybe another configurable property)